### PR TITLE
fix(redis source): fix high cpu on redis intermittent connection issues

### PR DESCRIPTION
**Pull request by hargut**
error scenario:
- vector starts and successfully initializes the redis source
- the connection experiences issues
- which lead to high CPU usage (>1 core)

the fix introduces a backoff at 100ms in case of a `RedisError::ErrorKind::IoError` to remediate the high CPU issue during described scenario

https://github.com/vectordotdev/vector/issues/16492

### CODE REVIEW

**Review by StephenWakely — APPROVED** [review]
Nice. Thanks for this. I'm wondering if an exponential backoff would be even more useful, but this will at least improve the situation.

**Comment by hargut** [comment]
I've updated it to an exponential backoff with a cap at 1000ms and fixed the fmt issue.

File: src/sources/redis/list.rs [hunk 1]
```diff
@@ -46,6 +61,11 @@ impl InputHandler {
     }
 }
 
+async fn backoff_exponential(exp: &u32) {
```
Comment (StephenWakely) [inline src/sources/redis/list.rs:64]: This doesn't need to be a reference.

```suggestion
async fn backoff_exponential(exp: u32) {
```

File: src/sources/redis/list.rs [hunk 2]
```diff
@@ -46,6 +61,11 @@ impl InputHandler {
...
+    let ms = if *exp <= 9 { 2_u64.pow(*exp) } else { 1000 };
```
Comment (StephenWakely) [inline src/sources/redis/list.rs:65]: Can we start at 32 and move up from there?

```suggestion
    let ms = if exp <= 4 { 2_u64.pow(exp + 5) } else { 1000 };
```

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Thanks for making that change, I think just a couple of minor tweaks and we should be good.

**Comment by hargut** [comment]
Hi @StephenWakely,

thank you for the review & suggestions. Added it to the PR.

**Pull request merged**